### ISSUE TITLE
File target archive compression goes out of control if log file is held open by other process

### ISSUE THREAD

**Issue by stylpe** [NLog/NLog#1314]
**Type** (choose one): Bug

**NLog version**: 4.1.0 and 4.3.0-beta3

**Platform**: .Net 4.5, Windows 10

**Current NLog config** (xml or C#, if relevant)

``` xml
  <targets>
    <default-wrapper xsi:type="AsyncWrapper"/>
    ...
    <target name="file" xsi:type="File" layout="${commonLayoutIndented}" 
            fileName="${configFolder}/MediaService.log" 
            createDirs="true" concurrentWrites="false" keepFileOpen="true"
            archiveNumbering="DateAndSequence" archiveDateFormat="yyyy.MM.dd"
            archiveAboveSize="52428800" maxArchiveFiles="10"
            archiveFileName="${configFolder}/MediaService.log.{#}.zip"
            enableArchiveFileCompression="true"
            encoding="UTF-8"/>
  </targets>
```

(${configFolder} is under C:/ProgramData/)

**Description**: We were experimenting with enableArchiveFileCompression. The above file target works wonderfully if the log files are left alone. We were also testing using tail from the Git for Windows installation (portable version also works, both available at https://git-scm.com/download/win) to watch our log file. When the rollover limit was hit, the log file stopped growing, and several empty archive files started rapidly being created and rolling over, very similar to what is described in #1028 (but we have concurrentWrites="false"). When we stopped the tail process, it seems to recover.

We tried updating to 4.3.0-beta3 since #1028 was fixed for that release, and with that version, it instead just stops logging, creates a single archive, and then nothing happens (no further log messages anywhere), until we stop tail, then it will continue as if nothing happened (spamming all the buffered log events so far to disk).

Tail is probably not the only program that triggers this, but I won't speculate on how tail is opening the file (or what NLog actually does when it rotates/archives the file). It's just a concrete program that can reproduce this.

Nothing special from the internal nlog log, for 4.1.0 we get nothing at all, and from 4.3.0-beta3 we get a single line: 2016-03-18 16:49:42.8018 Info Archiving ...\MediaService.log to zip-archive ...\MediaService.log.2016.03.18.0.zip

Possible workarounds: Don't use stupid tail from Git for Windows :P

On a more serious note, if we remove enableArchiveFileCompression, there are no problems for either version, and archiving and rotation works as expected (tail won't see the new log, but that's its problem)

**Comment by 304NotModified** [comment]
thanks for reporting! Also great that you have tested it in the latest beta :)

We use a lot of git, but personally I never used tail. What is it doing? Showing the most recent lines of the file? Does it modify the file (timestamp)?

**Comment by stylpe** [comment]
Ah, could've been more specific! Git for Windows comes with (at least portions of) Cygwin, so this is simply a Windows build of the Linux command line tool tail :)

**Comment by 304NotModified** [comment]
Brrr cygwin.  I'm powershell user ;) 

So any idea if tail edits the file (metadate) or locks it?

**Comment by stylpe** [comment]
Not for sure, only vague hypotheses. I doubt it edits the file or its attributes. I think it either locks it, or something weird is going on when NLog deletes it while rolling over (is this what it does btw? and not copy + truncate? haven't looked at the code), and tail somehow manages to "hold the spot", preventing NLog from creating the new file. The really odd thing is that it works fine without compressed archives :/

**Comment by stylpe** [comment]
I did some more investigating, and I think I was right about my second guess: tail keeps the file alive. I don't know exactly how, but it's easy to see it in action:

- run my application
- tail the log
- wait for it to try to archive
- stop my application (the log file will still exist, or at least is visible in Explorer)
- stop tail (the log file then disappears)

This explains why it works without compression (since it will just rename the log file in that case, while with compression it deletes and recreates it), and why 4.1.0 and 4.3.0 are different: 4.1.0 tries and fails to rollover for every message that is logged (I finally managed to get a stack trace for this by removing asyncwrapper and setting throwExceptions="true", it's a generic UnauthorizedAccessException: Access to the path is denied when opening the main log file). And for 4.3.0, looking at https://github.com/NLog/NLog/blob/v4.3.0-rc2/src/NLog/Targets/FileTarget.cs#L1177 there's the method DeleteAndWaitForFileDelete, which I bet will silently wait forever in this scenario. 

So, a potential solution is to rename the log file to some temp name just like when compression is not enabled, before deleting that.
